### PR TITLE
Feathers Avatar Upload Works

### PR DESCRIPTION
**Pull request by HexaField**
## Summary

- [x] generic feathers XHR uploader
- [x] location avatar upload works
- [x] admin avatar upload/view uses location modal
- [x] css works both in locaiton and admin
- [ ] readyplayerme uses new upload & works properly

refactors the large file upload XHR function to be service agnostic, and fixes the problems with avatar uploading which was due to not properly waiting for the request to complete.

## Checklist
- [ ] Pre-push checks pass `npm run check`
  - [ ] Linter passing via `npm run lint`
  - [ ] Unit & Integration tests passing via `npm run test:packages`
  - [ ] Docker build process passing via `npm run build-client`
- [ ] If this PR is still a WIP, convert to a draft 
- [ ] When this PR is ready, mark it as "Ready for review"
- [ ] Changes have been manually QA'd
- [ ] Changes reviewed by at least 2 approved reviewers

## References

closes https://github.com/XRFoundation/XREngine/issues/5577

## QA Steps

1. `git checkout pr_branch_name`
2. `npm install`
3. `npm run dev-reinit`
4. `npm run dev`

List any additional steps required to QA the changes of this PR, as well as any supplemental images or videos.

## Reviewers

_Reviewers for this PR_

### CODE REVIEW

**Comment by zulqarnainhanif** [comment]
@HexaField Please look into the `client-core/src/admin/components/Avatars/AvatarCreate.tsx` file as well it uses the similar code that might need fixing. Both Avatar components can use helper functions a lot/can be cleaned up.

**Comment by zulqarnainhanif** [comment]
@HexaField can you help with the below view and thumbnail created against the avatar in admin panel as well.

Screenshot 2022-04-07 at 3 09 36 PM

**Comment by zulqarnainhanif** [comment]
newly added avatar from admin panel have no preview thumbnail.

Screenshot 2022-04-07 at 3 17 39 PM

**Comment by HexaField** [comment]
@zulqarnainhanif hmm yes, the admin panel will need to initialize the engine for it to be able to use the location UI, lets open this in a new issue

File: packages/client-core/src/util/upload.tsx [hunk 1]
```diff
@@ -14,139 +13,66 @@ const serverURL =
...
+
+    request.addEventListener('readystatechange', (e) => {
+      // console.log('readystatechange', e, request.readyState)
+      if (request.readyState === XMLHttpRequest.DONE) {
+        const status = request.status
+        if (status === 0 || (status >= 200 && status < 400)) {
+          console.log('The request has been completed successfully')
+          resolve()
+        } else {
+          console.log('Oh no! There has been an error with the request!')
```
Comment (barankyle) [inline packages/client-core/src/util/upload.tsx:54]: The error could be logged here to make debugging easier.

File: packages/engine/src/input/schema/ClientInputSchema.ts [hunk 2]
```diff
@@ -476,6 +476,7 @@ export const handleKey = (event: KeyboardEvent): any => {
   }
   // const mappedKey = Engine.inputState.schema.keyboardInputMap[];
   const key = event.code
+  console.log(key)
```
Comment (barankyle) [inline packages/engine/src/input/schema/ClientInputSchema.ts:479]: Unnecessary console.log?

File: packages/server-core/src/user/avatar/avatar-helper.ts [hunk 3]
```diff
@@ -100,6 +100,7 @@ export const getAvatarFromStaticResources = async function (app: Application, na
     },
     isInternal: true
   })
+  console.log('avatarQueryResult', avatarQueryResult)
```
Comment (barankyle) [inline packages/server-core/src/user/avatar/avatar-helper.ts:103]: Remove?

File: packages/client-core/src/user/components/UserMenu/menus/AvatarSelectMenu.tsx [hunk 4]
```diff
@@ -457,8 +460,8 @@ export const AvatarSelectMenu = (props: Props) => {
             type="button"
             className={styles.uploadBtn}
             onClick={uploadAvatar}
-            style={{ cursor: !fileSelected || !!error ? 'not-allowed' : 'pointer' }}
-            disabled={!fileSelected || !!error}
+            style={{ cursor: uploadButtonEnabled ? 'pointer' : 'not-allowed' }}
```
Comment (barankyle) [inline packages/client-core/src/user/components/UserMenu/menus/AvatarSelectMenu.tsx:463]: Should the upload button be disabled once a file is selected? I can see someone mis-clicking and wanting to change the selection.

**Review by barankyle — CHANGES_REQUESTED** [review]
With local storage, uploaded avatars' filenames are coming out as e.g. avatars/public/1649701290717CyberbotSilver.glb. Is the number supposed to be a sub-directory with the .glb inside that sub-directory?

While the uploaded file has an extension of .glb, its file type is 'image'. This might be why, when I try to change to them in a scene, nothing gets rendered.

Is the uploader supposed to create a default thumbnail and/or have some way to upload a thumbnail?

I tried uploading from this URL: https://resources-dev.theoverlay.io/avatars/public/Razer3.glb and the resulting file had the name 1649701869997undefined . When I opened the uploader again and pasted the same source URL in, there's an error about the URL being invalid.

File: packages/client-core/src/user/components/UserMenu/menus/AvatarSelectMenu.tsx (hunk 4 above)
Comment (HexaField) [reply packages/client-core/src/user/components/UserMenu/menus/AvatarSelectMenu.tsx]: the admin panel now uses the new logic and is working properly

**Comment by HexaField** [comment]
> @HexaField the view for pre-uploaded avatars shows up empty in admin panel.
> 
> Screenshot 2022-04-12 at 1 30 23 PM

this seems to be because an avatar is uploaded with no name, which shouldnt happen, it appears this is from something else

**Comment by HexaField** [comment]
> thumbnail generation looks off
> 
> Screenshot 2022-04-12 at 1 37 10 PM

this is because the avatar was not positioned when the upload button was pressed. this funcitonality is good enough for now, we can improve this workflow when we implement the character creator

**Review by speigg — APPROVED** [review]

**Pull request merged**